temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowClientInternal.java

```diff
@@ -62,15 +65,18 @@
   private final Scope metricsScope;
 
   /**
-   * Creates client that connects to an instance of the Temporal Service.
+   * Creates client that connects to an instance of the Temporal Service. Cannot be used from within
+   * workflow code.
    *
    * @param service client to the Temporal Service endpoint.
    * @param options Options (like {@link io.temporal.common.converter.DataConverter} override) for
    *     configuring client.
    */
   public static WorkflowClient newInstance(
       WorkflowServiceStubs service, WorkflowClientOptions options) {
-    return new WorkflowClientInternal(service, options);
+    enforceNonWorkflowThread();
+    return WorkflowThreadMarker.protectFromWorkflowThread(
```

Comment:
We should consider adding an option that would disable this enforcement. If users are already running "bad" workflows in production, then there should be a way to disable this feature until workflows are refactored/updated, otherwise we are forcing users to stay on older SDK versions.